File: THCS/2023_2024/HaNoi_2324/Bai2/UOCCHUNG.py
import math

# Hàm tìm ước của một số
def find_divisors(n):
    divisors = []
    for i in range(1, int(math.sqrt(n)) + 1):
        if n % i == 0:
            divisors.append(i)
            if i != n // i:
                divisors.append(n // i)
    return sorted(divisors)

# Đọc dữ liệu từ tệp UOCCHUNG.INP
with open("UOCCHUNG.INP", "r") as f:
    A, B = map(int, f.readline().split())

# Tìm ước chung lớn nhất (GCD)
gcd_value = math.gcd(A, B)

# Tìm các ước của GCD
divisors = find_divisors(gcd_value)

# Lấy ước lớn thứ hai
second_largest_divisor = divisors[-2] if len(divisors) > 1 else divisors[0]

# Ghi kết quả vào tệp UOCCHUNG.OUT
with open("UOCCHUNG.OUT", "w") as f:
    f.write(str(second_largest_divisor))

File: THCS/2023_2024/HaNoi_2324/Bai2/UOCCHUNG.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

// Hàm tìm ước của một số
vector<long long> findDivisors(long long n) {
    vector<long long> divisors;
    for (long long i = 1; i <= sqrt(n); ++i) {
        if (n % i == 0) {
            divisors.push_back(i);
            if (i != n / i) {
                divisors.push_back(n / i);
            }
        }
    }
    sort(divisors.begin(), divisors.end());
    return divisors;
}

// Hàm tìm ước chung lớn nhất (GCD) sử dụng thuật toán Euclid
long long gcd(long long a, long long b) {
    while (b != 0) {
        long long temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int main() {
    ifstream input("UOCCHUNG.INP");
    ofstream output("UOCCHUNG.OUT");
    
    long long A, B;
    input >> A >> B;

    // Tìm ước chung lớn nhất (GCD)
    long long gcdValue = gcd(A, B);

    // Tìm các ước của GCD
    vector<long long> divisors = findDivisors(gcdValue);

    // Lấy ước lớn thứ hai
    long long secondLargestDivisor = (divisors.size() > 1) ? divisors[divisors.size() - 2] : divisors[0];

    // Ghi kết quả vào tệp
    output << secondLargestDivisor << endl;

    input.close();
    output.close();

    return 0;
}
